File: src/services/queue_service.py
import json
from datetime import datetime
from src.config.db import get_db

class QueueService:
    def publish(self, event, payload, db_conn=None):
        try:
            if db_conn:
                self._insert(db_conn, event, payload)
            else:
                with get_db() as db:
                    self._insert(db, event, payload)
            print(f"[QUEUE] Event Published: {event}")
        except Exception as e:
            print(f"[QUEUE] Failed to publish event {event}: {str(e)}")
            
    def _insert(self, db, event, payload):
        now = datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
        db.execute('''
            INSERT INTO notification_queue (event_type, payload, status, next_attempt_at)
            VALUES (?, ?, ?, ?)
        ''', (event, json.dumps(payload), 'pending', now))

queue_service = QueueService()

File: src/services/queueService.js
const db = require('../config/db');

/**
 * Internal Queue Service for Event Publishing
 * This implements a database-backed queue for the MVP Operational Layer.
 */
class QueueService {
    /**
     * Publish an event to the notification queue
     * @param {string} event - The event type (e.g., 'block.created')
     * @param {object} payload - The event data
     * @param {object} trx - Optional transaction object
     */
    async publish(event, payload, trx = null) {
        const t = trx || db;
        try {
            await t('notification_queue').insert({
                event_type: event,
                payload: JSON.stringify(payload),
                status: 'pending',
                next_attempt_at: db.fn.now()
            });
            console.log(`[QUEUE] Event Published: ${event}`);
        } catch (error) {
            console.error(`[QUEUE] Failed to publish event ${event}:`, error.message);
        }
    }
}

module.exports = new QueueService();
